CONSOLE-3383: Minimal Dynamic Plugin for the CronTab CRD (MVP)

jhadvig opened the pull request:
This PR contains a MVP for crontab dynamic plugin. It's adding:
- custom list page with custom action provider (action button / kebab)
- yaml template for crontab CR
- i18n support
- README

List page screen:
<img width="1916" alt="Screenshot 2023-05-17 at 11 27 56" src="https://github.com/openshift/crontab-plugin/assets/1668218/40603b99-bcf3-4fe7-8461-f1196a294248">


Create page:
<img width="1914" alt="Screenshot 2023-05-17 at 11 27 45" src="https://github.com/openshift/crontab-plugin/assets/1668218/6c1f677d-798e-4abc-b058-a25b629ab1d8">


/assign @spadgett 

FYI: I've preserved whole commit history. 
CCing: @raspbeep

spadgett commented on Dockerfile:
```diff
@@ -0,0 +1,10 @@
+FROM docker.io/library/node:16 AS build
```
Let's update the Dockerfile to what we use in the console plugin template

spadgett commented on charts/crontab-plugin/Chart.yaml:
```diff
@@ -0,0 +1,5 @@
+apiVersion: v2
+name: crontab-plugin
+description: A Helm chart for Kubernetes
```
Let's give this a better description.

spadgett commented on console-extensions.json:
```diff
@@ -0,0 +1,69 @@
+    }
+  },
+  {
+    "type": "console.navigation/resource-ns",
```
We should make sure the nav extension is only enabled when the feature flag is set.

spadgett commented on i18n-scripts/lexers.js:
```diff
@@ -0,0 +1,30 @@
+const EventEmitter = require('events');
```
We should consider moving this into the template.

jhadvig replied:
not sure what template you mean?

spadgett replied:
openshift/console-plugin-template

jhadvig replied:
will create a story for that

spadgett commented on locales/en/plugin__crontab-plugin.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "Actions": "",
```
Should the message keys be empty like this?

jhadvig replied:
no it does not.. hmm not sure whats happening since I have `useKeysAsDefaultValue: true,` in my `i18next-parser.config.js`

jhadvig replied:
also `yarn i18n` scripts passes without any issue and creates the locale file :-/

spadgett commented on oc-manifest.yaml:
```diff
@@ -0,0 +1,152 @@
+apiVersion: v1
```
Do we need both the manifest and the Helm chart? I would prefer just having one.

jhadvig replied:
The motivation here was that not everybody is using Helm, and would rather prefer using `oc`

jhadvig replied:
removed the `oc-manifest.yaml` and updated the README

spadgett commented on package.json:
```diff
@@ -0,0 +1,83 @@
+    "name": "crontab-plugin",
+    "version": "0.0.1",
+    "displayName": "OpenShift CronTab Plugin",
+    "description": "Sample plugin for OpenShift Console. This plugin adds basic functionalities to the OpenShift Console that allows users to manage CronTab CRD.",
```
```suggestion
    "description": "Sample plugin for the OpenShift Console. This plugin adds basic functionality that allows users to manage CronTab custom resources.",
```

spadgett commented on package.json:
```diff
@@ -0,0 +1,83 @@
+  "consolePlugin": {
+    "name": "crontab-plugin",
+    "version": "0.0.1",
+    "displayName": "OpenShift CronTab Plugin",
```
```suggestion
    "displayName": "Sample CronTab Plugin",
```

spadgett commented on src/models/CronTabModel.ts:
```diff
@@ -0,0 +1,25 @@
+  crd: true,
+};
+
+export type CronTabKind = K8sResourceCommon & {
+  spec?: {
+    [key: string]: any;
+  };
+  status?: { [key: string]: any };
+};
```
It would be better to define a more strict type here.

spadgett commented on src/models/CronTabModel.ts:
```diff
@@ -0,0 +1,25 @@
+  status?: { [key: string]: any };
+};
+
+export const cronTabModelRef = modelToRef(CronTabModel);
```
We intended to remove the concept of resource references from the SDK... Where is this needed?

jhadvig replied:
this was used when manually building the URLs

jhadvig replied:
So Im not sure how to avoid building the url manually at this point.
I start to think this is a tech debt on our side and we should expose a function which will help plugins creators with crafting URLs.
Thoughts?

spadgett replied:
Yeah, I think we need an SDK method for building resource URLs. Then if we ever need to change how the URLs are built (e.g. if we remove the API resource version from the URL) plugins aren't broken.

spadgett commented on src/models/CronTabModel.ts:
```diff
@@ -0,0 +1,25 @@
+import { K8sResourceCommon } from '@openshift-console/dynamic-plugin-sdk';
+import { K8sModel } from '@openshift-console/dynamic-plugin-sdk/lib/api/common-types';
+
+export const CronTabModel: K8sModel = {
```
I think this should be defined using the  model metadata extension. We shouldn't need to reference the model directly in the API IIRC, so we hopefully we can remove this.

https://github.com/openshift/console/blob/master/frontend/packages/console-dynamic-plugin-sdk/docs/console-extensions.md#consolemodel-metadata

spadgett replied:
This comment still stands. I don't think we should be creating a K8sModel resource, even if we have to build a reference somewhere.

spadgett commented on src/utils/utils.ts:
```diff
@@ -0,0 +1,10 @@
+import { K8sModel } from '@openshift-console/dynamic-plugin-sdk/lib/api/common-types';
+
+const modelToRef = (obj: K8sModel) => `${obj.apiGroup}~${obj.apiVersion}~${obj.kind}`;
+const modelToGroupVersionKind = (obj: K8sModel) => ({
+  version: obj.apiVersion,
+  kind: obj.kind,
+  group: obj.apiGroup,
+});
```
It would be good to try to remove these. We don't want plugins to need to do these kinds of transformations.

spadgett commented on src/views/CronTabList/CronTabList.tsx:
```diff
@@ -0,0 +1,157 @@
+  return (
+    <>
+      <ListPageHeader title={t('CronTab')}>
+        <ListPageCreate groupVersionKind={modelToRef(CronTabModel)}>Create CronTab</ListPageCreate>
```
Should `groupVersionKind` be the object, not the reference string? If that's not the case, we need to open a bug against the SDK.

jhadvig replied:
Unfortunately `groupVersionKind` is type of [GroupVersionKind](https://github.com/openshift/console/blob/master/frontend/packages/console-dynamic-plugin-sdk/src/api/dynamic-core-api.ts#L177) which is string.
We need to update to K8sGroupVersionKind which is an object

jhadvig replied:
Created https://issues.redhat.com/browse/OCPBUGS-13808

spadgett commented on src/views/CronTabList/CronTabList.tsx:
```diff
@@ -0,0 +1,157 @@
+const cronTabListRow: React.FC<RowProps<CronTabKind>> = ({ obj, activeColumnIDs }) => {
+  return (
+    <>
+      <TableData id="name" activeColumnIDs={activeColumnIDs} className="pf-m-width-15">
```
Do we need these width classes? By default the columns should all be the same width?

jhadvig replied:
removed

spadgett commented on src/views/hooks/useCronTabActionsProvider.tsx:
```diff
@@ -0,0 +1,31 @@
+        label: t('Edit'),
+        cta: () =>
+          history.push(
+            `/k8s/ns/${cronTab.metadata.namespace}/${cronTabModelRef}/${cronTab.metadata.name}/yaml`,
```
Avoid manually building URLs if at all possible.

spadgett reviewed (commented):
Thanks @jhadvig

jhadvig commented on src/views/components/CronTabActions/CronTabActions.tsx:
```diff
@@ -0,0 +1,67 @@
+  const [isOpen, setIsOpen] = React.useState<boolean>(false);
+  const [actions /*, onLazyOpen*/] = useCronTabActionsProvider(cronTab);
+
+  const handleClick = (action: Action) => {
+    if (typeof action?.cta === 'function') {
+      action.cta();
+    }
+    setIsOpen(false);
+  };
```
```suggestion
  const handleClick = (action: Action) => {
    if (typeof action?.cta === 'function') {
      action.cta();
    } else {
      console.error('CronTabActions: action.cta is not a function');
    }
    setIsOpen(false);
  };
```

jhadvig replied:
added following condition

jhadvig commented on src/views/components/CronTabActions/CronTabActions.scss:
```diff
@@ -0,0 +1,11 @@
+.console-cron-tab-actions {
+    ul {
+      list-style: none;
+      padding-left: 0;
+    }
```
not needed... removed it

jhadvig commented on src/views/components/CronTabActions/CronTabActions.scss:
```diff
@@ -0,0 +1,11 @@
+      padding-left: 0;
+    }
+  
+    .pf-c-dropdown__menu {
```
removed

jhadvig commented on src/views/CronTabList/CronTabList.tsx:
```diff
@@ -0,0 +1,157 @@
+      <TableData
+        id="actions"
+        activeColumnIDs={activeColumnIDs}
+        className="dropdown-kebab-pf pf-c-table__action"
```
removed

jhadvig commented:
@spadgett addressed most of the comments.
Regarding building URL manually and having models part of the codebase, I think we need to address those as a techdebt into the SDK.
Will create a follow up stories for those.

spadgett commented on .gitignore:
```diff
@@ -0,0 +1,8 @@
+node_modules/
+dist/
+.devcontainer/dev.env
+integration-tests/videos
+integration-tests/screenshots
+integration-tests/.DS_Store
```
We don't have any integration tests yet, correct?

spadgett commented on README.md:
```diff
@@ -0,0 +1,58 @@
+
+It requires OpenShift 4.11.
+
+The CronTab Dynamic Plugin creates new menu entry, routes, list page, details page and a
+default YAML template for a CRD.
```
```suggestion
The CronTab Dynamic Plugin creates a new navigation entry, list page, details page, and YAML template for the CRD.
```

spadgett commented on charts/crontab-plugin/Chart.yaml:
```diff
@@ -0,0 +1,5 @@
+apiVersion: v2
+name: crontab-plugin
+description: A Helm chart for CronTab Dynamic Plugin used in OpenShift Console
```
```suggestion
description: A Helm chart for the sample CronTab Dynamic Plugin used in OpenShift Console.
```

spadgett commented on i18n-scripts/common.js:
Are we sure we need all of these i18n scripts? This is a lot of code to be copying around.

If they're really needed, we should make this available as a package that plugins can use. But I'm surprised we need this much custom code to use i18next-parser for a simple project.

jhadvig replied:
So I was trying to decouple it but the default values havent been generated for the strings, for some reason.

spadgett commented on src/views/CronTabList/CronTabRowActions.tsx:
```diff
@@ -0,0 +1,46 @@
+
+  const onEditCronTab = () => {
+    const cta = {
+      href: `/k8s/ns/${obj.metadata.namespace}/${cronTabModelRef}/${obj.metadata.name}/yaml`,
```
I'm inclined to remove the actions from the initial PR until we can fix this properly. The problem is people will copy this sample code into their own plugins, and even if we update the sample, those real plugins won't get updated.

I'd suggest we get something working for the current released OpenShift versions and make sure they're in an appropriate release-* branch, then we can come back and add actions when the SDK functions we need are available.

jhadvig replied:
Makes sense. On the other side folks will check how other plugins are doing (constructing the URL manually) and use that.

But I like it more then having it here... will remove 👍

spadgett commented on src/views/components/CronTabActions/CronTabActions.tsx:
```diff
@@ -0,0 +1,66 @@
+      }
+      dropdownItems={actions?.map((action) => (
+        <DropdownItem
+          data-test-id={action?.id}
```
Remove data-test-id if we don't have any test code

jhadvig replied:
removed

spadgett commented on src/views/CronTabList/CronTabList.tsx:
```diff
@@ -0,0 +1,151 @@
+    <>
+      <TableData id="name" activeColumnIDs={activeColumnIDs} >
+        <ResourceLink
+          groupVersionKind={modelToGroupVersionKind(CronTabModel)}
```
Just use

```
    groupVersionKind: {
      group: 'stable.example.com',
      kind: 'CronTab',
      version: 'v1',
    }
```

and remove the modelTo helper

jhadvig replied:
done

spadgett reviewed (commented):
Thanks, @jhadvig

jhadvig commented:
@spadgett ping

spadgett commented on src/utils/utils.ts:
```diff
@@ -0,0 +1,9 @@
+import { K8sModel } from '@openshift-console/dynamic-plugin-sdk/lib/api/common-types';
+
+const modelToGroupVersionKind = (obj: K8sModel) => ({
```
Is this still used?

jhadvig replied:
removed

spadgett commented on src/views/CronTabList/CronTabList.tsx:
```diff
@@ -0,0 +1,155 @@
+  return (
+    <>
+      <ListPageHeader title={t('CronTab')}>
+        {/* groupVeersionKind should be taking object but there is dicrepancy in the prop types
+        https://issues.redhat.com/browse/OCPBUGS-13808 should be updateing the types */}
```
```suggestion
        {/* groupVeersionKind should be take an object but there is discrepancy in the prop types
        https://issues.redhat.com/browse/OCPBUGS-13808 will update the types */}
```

spadgett commented on src/views/CronTabList/CronTabList.tsx:
```diff
@@ -0,0 +1,155 @@
+      <ListPageHeader title={t('CronTab')}>
+        {/* groupVeersionKind should be taking object but there is dicrepancy in the prop types
+        https://issues.redhat.com/browse/OCPBUGS-13808 should be updateing the types */}
+        <ListPageCreate groupVersionKind={`stable.example.com~v1~CronTab`}>Create CronTab</ListPageCreate>
```
```suggestion
        <ListPageCreate groupVersionKind="stable.example.com~v1~CronTab">Create CronTab</ListPageCreate>
```

spadgett commented on src/views/CronTabList/CronTabList.tsx:
```diff
@@ -0,0 +1,155 @@
+    <>
+      <TableData id="name" activeColumnIDs={activeColumnIDs} >
+        <ResourceLink
+          groupVersionKind={{
```
nit: I'd pull this out into a const somewhere so we don't repeat it in several places

spadgett commented on src/views/CronTabList/CronTabRowActions.tsx:
```diff
@@ -0,0 +1,29 @@
+      toggle={<KebabToggle onToggle={setIsDropdownOpen} id="toggle-id-disk" />}
+      isOpen={isDropdownOpen}
+      isPlain
+      dropdownItems={[]}
```
If there are no items, I would leave the dropdown out of the PR for now.

jhadvig replied:
wanted to preserve, if it would be used as a starting point for a new plugin, but can remove

spadgett commented on src/views/hooks/useCronTabActionsProvider.tsx:
```diff
@@ -0,0 +1,18 @@
+
+  const actions = React.useMemo(
+    () => [],
+    [],
```
If there are no actions, I'd leave this out. We can keep this in a separate branch so we still have the code.

jhadvig replied:
removed

spadgett commented on src/views/CronTabList/CronTabRowActions.tsx:
I would take this entire component out from the PR if we have no actions.

We can add the edit crontab modal in a follow up and add this back.

jhadvig replied:
removed

jhadvig commented:
@spadgett comments addressed :)

spadgett reviewed (commented):
LGTM, can we clean up the commit history a bit before merging? It doesn't have to be a single commit, but would be nice to squash things to remove "WIP" and "addressing comments" type commits.

jhadvig commented:
For sure ! was just waiting for LGTM 😅

spadgett commented on src/views/components/CronTabActions/CronTabActions.tsx:
Not sure if I missed on previous review, but we should remove until we have actions.

jhadvig commented:
@spadgett comment addressed. PTAL

spadgett reviewed (approved):
/lgtm
